here-naksha-app-service/src/test/java/com/here/naksha/app/service/WriteFeaturesAtomicityTest.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.app.service;
+
+import static com.here.naksha.app.common.CommonApiTestSetup.setupSpaceAndRelatedResources;
+import static com.here.naksha.app.common.ResponseAssertions.assertThat;
+import static com.here.naksha.app.common.TestUtil.loadFileOrFail;
+
+import com.here.naksha.app.common.ApiTest;
+import com.here.naksha.app.common.CommonApiTestSetup;
+import com.here.naksha.app.common.NakshaTestWebClient;
+import com.here.naksha.app.common.ResponseAssertions;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpResponse;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+class WriteFeaturesAtomicityTest extends ApiTest {
+
+  private static final NakshaTestWebClient nakshaClient = new NakshaTestWebClient();
+  private static final String SPACE_ID = "write_features_atomicity_test_space";
+
+  public WriteFeaturesAtomicityTest() {
+    super(nakshaClient);
+  }
+
+  @BeforeAll
+  static void prepareEnv() {
+    setupSpaceAndRelatedResources(nakshaClient, "WriteFeaturesAtomicity/setup");
+  }
+
+  @Test
+  void tc_1101_partialCreateShouldFail() throws URISyntaxException, IOException, InterruptedException {
+    // Given: multiple features to save, some of which are invalid
+    List<String> validFeatureIds = List.of("tc_1101_feature_1", "tc_1101_feature_2");
+    String createFeaturesJson = loadFileOrFail("WriteFeaturesAtomicity/TC1101_partialCreateShouldFail/create_features.json");
+
+    // When: saving these features
+    String streamId = UUID.randomUUID().toString();
+    HttpResponse<String> saveResp = nakshaClient.post("hub/spaces/" + SPACE_ID + "/features", createFeaturesJson, streamId);
+
+    // Then: response indicates failure
+    ResponseAssertions.assertThat(saveResp)
+        .hasStatus(500)
+        .hasStreamIdHeader(streamId);
+
+    // And: none of the features got saved
+    String idsQuery = "?id=" + String.join("&id=", validFeatureIds);
+    HttpResponse<String> getResp = nakshaClient.get("hub/spaces/" + SPACE_ID + "/features" + idsQuery, streamId);
+    ResponseAssertions.assertThat(getResp)
+        .hasStatus(200)
+        .hasStreamIdHeader(streamId)
+        .hasJsonBody(loadFileOrFail("WriteFeaturesAtomicity/TC1101_partialCreateShouldFail/empty_response.json"));
+  }
+
+  @Test
+  void tc_1102_partialUpdateShouldFail() throws URISyntaxException, IOException, InterruptedException {
+    // Given: multiple features to save (all valid)
+    String streamId = UUID.randomUUID().toString();
+    List<String> featureIds = List.of("tc_1102_feature_1", "tc_1102_feature_2");
+    String createFeatureJson = loadFileOrFail("WriteFeaturesAtomicity/TC1102_partialUpdateShouldFail/create_features.json");
+    nakshaClient.post("hub/spaces/" + SPACE_ID + "/features", createFeatureJson, streamId);
```

Comment:
nth: like the setups, adding assert 200 status code here might help spot unexpected problem sooner

Reply:
will do :)